relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+    hasher: ahash::RandomState,
+}
+
+impl Inner {
+    pub fn new(config: Config) -> Self {
+        let bucket_interval = config.bucket_interval.max(1);
+        // Extend the configured time slots with the delay (in slots), to make sure there is
+        // enough space to satisfy the delay.
+        //
+        // We cannot just reserve space for enough partitions to satisfy the delay, because we have
+        // no control over which partition a bucket is assigned to, so we have to prepare for the
+        // 'worst' case, and that is full extra time slots.
+        let num_time_slots = config.num_time_slots.max(1) + config.delay.div_ceil(bucket_interval);
+        let num_partitions = config.num_partitions.max(1);
+
+        let mut slots = Vec::new();
+        for _ in 0..num_time_slots {
+            for partition_key in 0..num_partitions {
+                slots.push(Slot {
+                    partition_key,
+                    ..Default::default()
+                });
+            }
+        }
+
+        let slot_diff = SlotDiff {
+            min: config
+                .max_secs_in_past
+                .map_or(u64::MAX, |v| v.div_ceil(u64::from(bucket_interval))),
+            max: config
+                .max_secs_in_future
+                .map_or(u64::MAX, |v| v.div_ceil(u64::from(bucket_interval))),
+        };
+
+        let total_slots = slots.len();
+        Self {
+            slots: VecDeque::from(slots),
+            num_partitions,
+            head: config.start.as_secs() / u64::from(bucket_interval) * u64::from(num_partitions),
+            bucket_interval,
+            delay: config.delay,
+            stats: stats::Total::default(),
+            limits: stats::Limits {
+                max_total: config.max_total_bucket_bytes.unwrap_or(u64::MAX),
+                // Break down the maximum project cost to a maximum cost per partition.
+                max_partition_project: config
+                    .max_project_key_bucket_bytes
+                    .map(|c| c.div_ceil(total_slots as u64))
+                    .unwrap_or(u64::MAX),
+            },
+            slot_diff,
+            partition_by: FlushBatching::Partition,
+            hasher: build_hasher(),
+        }
+    }
+
+    /// Returns the configured bucket interval.
+    pub fn bucket_interval(&self) -> u32 {
+        self.bucket_interval
+    }
+
+    /// Returns total aggregator stats.
+    pub fn stats(&self) -> Stats {
+        Stats {
+            count: self.stats.count,
+            count_by_namespace: self.stats.count_by_namespace,
+            cost: self.stats.count,
+            cost_by_namespace: self.stats.cost_by_namespace,
+        }
+    }
+
+    /// Returns `true` if the aggregator contains any metric buckets.
+    pub fn is_empty(&self) -> bool {
+        self.stats.cost == 0
+    }
+
+    /// Returns the time as a duration since epoch when the next flush is supposed to happen.
+    pub fn next_flush(&self) -> Duration {
```

Comment:
No, since this is no longer `full seconds`, it returns the next timestamp of when the partition is to be flushed. The Duration can represent that and deals with the division down to ns precision.